## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [2]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head(5)

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [3]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [4]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [5]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [6]:
# What three item categories had the most entries?
count_category_df = df.groupby("category").agg(category_count = ("category", "count"))
top_category_df_count = count_category_df.sort_values("category_count", ascending=False).head(1)
top_category_df_count

category_count
category                   
consumables           23538

In [7]:
# For the category with the most entries, which subcategory had the most entries?
top_category_item_df =  df.join(top_category_df_count, how="inner", on="category")
top_category_item_df.groupby(["category","subcategory"]).agg(subcategory_count = ("subcategory", "count")).sort_values("subcategory_count", ascending=False).head(1)


subcategory_count
category    subcategory                         
consumables bathroom supplies               6424

In [8]:
# Which five clients had the most entries in the data?
df_top5_clients = df.groupby("client_id").agg(entry_count = ("client_id", "count")).sort_values("entry_count", ascending=False).head(5)
df_top5_clients

entry_count
client_id             
33615              220
66037              211
46820              209
38378              207
24741              207

In [9]:
# Store the client ids of those top 5 clients in a list.
lst_clientid = [x for x in df_top5_clients.index]
lst_clientid

[33615, 66037, 46820, 38378, 24741]

In [10]:
# How many total units (the qty column) did the client with the most entries order order?
top_client_item_df = df.loc[df["client_id"]==lst_clientid[0],:]
order_qty_df =  top_client_item_df.groupby("order_id").agg(nos_of_order = ("order_id","count"), total_units=("qty","sum"))
order_qty_df.sort_values(["nos_of_order","total_units"], ascending=False,inplace=True)
order_qty_df.head(1)


nos_of_order  total_units
order_id                           
7417275             10        18356

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [11]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
#df.head(5)
client_specific_df = df.copy()
client_specific_df["subtotal"] = client_specific_df["unit_price"] * client_specific_df["qty"]
client_specific_df.head(5)

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [12]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
client_specific_df["total_weight"] = client_specific_df["unit_weight"] * client_specific_df["qty"]
client_specific_df["shipping_price"] = round( client_specific_df["total_weight"].apply( lambda weight: (weight * 7.00) if  weight > 50.00 else weight * 10.00 ) ,2)
client_specific_df.head(5)


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number   subtotal  total_weight  \
0     762.71         7.50  105            1  115164.00        787.50   
1      15.09         1.49   21            0     523.95         31.29   
2       7.86         1.68   39            6     527.28         65.52   
3      24.85         1.23   29            3    1056.18         35.67   
4     108.17        46.43   20            1    3902.00        928.60   

   shipping_price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20  

[5 rows x 21 columns]

In [13]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
client_specific_df["total_tax"] = round(((client_specific_df["subtotal"] + client_specific_df["shipping_price"])*9.25/100),2)
client_specific_df["total_price"] = round(client_specific_df["subtotal"] + client_specific_df["shipping_price"] + client_specific_df["total_tax"],2) 
client_specific_df.head(5)


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   subtotal  total_weight  shipping_price  total_tax  \
0            1  115164.00        787.50         5512.50   11162.58   
1            0     523.95         31.29          312.90      77.41   
2            6     527.28         65.52          458.64      91.20   
3            3    1056.18         35.67          356.70     130.69   
4            1    3902.00        928.60         6500.20     962.20   

   total_price  
0    131839.08  
1       914.26  
2      1077.12  
3      1543.57  
4     11364.40  

[5 rows x 23 columns]

In [14]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
client_specific_df["line_cost"]  = round((client_specific_df["unit_cost"] * client_specific_df["qty"]) + client_specific_df["shipping_price"],2)
client_specific_df.head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

    subtotal  total_weight  shipping_price  total_tax  total_price  line_cost  
0  115164.00        787.50         5512.50   11162.58    131839.08   85597.05  
1     523.95         31.29          312.90      77.41       914.26     629.79  
2     527.28         65.52          458.64      91.20      1077.12     765.18  

[3 rows x 24 columns]

In [15]:
# Create a column for the profit of each line using line cost and line price
client_specific_df["profit"]  = client_specific_df["total_price"] - client_specific_df["line_cost"]
client_specific_df.head(5)


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_weight  qty line_number   subtotal  \
0          17        2023  ...        7.50  105           1  115164.00   
1          20        2023  ...        1.49   21           0     523.95   
2           4        2023  ...        1.68   39           6     527.28   
3          17        2023  ...        1.23   29           3    1056.18   
4          21        2023  ...       46.43   20           1    3902.00   

   total_weight  shipping_price  total_tax  total_price  line_cost    profit  
0        787.50         5512.50   11162.58    131839.08   85597.05  46242.03  
1         31.29          312.90      77.41       914.26     629.79    284.47  
2         65.52          458.64      91.20      1077.12     765.18    311.94  
3         35.67          356.70     130.69      1543.57    1077.35    466.22  
4        928.60         6500.20     962.20     11364.40    8663.60   2700.80  

[5 rows x 25 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [16]:
# Check your work using the totals above
def getTotal(x):
    return round(x.sum(),2)

order3_df = client_specific_df.groupby(["order_id"]).agg(order_price=("total_price", getTotal))

print(f"Total Price for order 2742071: ${order3_df.loc[2742071,"order_price"]}\n")
print(f"Total Price for order 2173913: ${order3_df.loc[2173913,"order_price"]}\n")
print(f"Total Price for order 6128929: ${order3_df.loc[6128929,"order_price"]}\n")


Total Price for order 2742071: $152811.89

Total Price for order 2173913: $162388.71

Total Price for order 6128929: $923441.25



## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [17]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.

df_clientid = pd.DataFrame(lst_clientid, columns=["client_id"])

top_client5_item_df = pd.merge(client_specific_df,df_clientid, on="client_id")
top_client5_total_qty_df = top_client5_item_df.groupby("client_id").agg(total_quantity = ("qty","sum"))
top_client5_total_qty_df.head(5)
top_client5_item_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1054 entries, 0 to 1053
Data columns (total 25 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   first           1054 non-null   object 
 1   last            1054 non-null   object 
 2   job             1054 non-null   object 
 3   phone           1054 non-null   object 
 4   email           1054 non-null   object 
 5   client_id       1054 non-null   int64  
 6   order_id        1054 non-null   int64  
 7   order_date      1054 non-null   object 
 8   order_week      1054 non-null   int64  
 9   order_year      1054 non-null   int64  
 10  item_id         1054 non-null   object 
 11  category        1054 non-null   object 
 12  subcategory     1054 non-null   object 
 13  unit_price      1054 non-null   float64
 14  unit_cost       1054 non-null   float64
 15  unit_weight     1054 non-null   float64
 16  qty             1054 non-null   int64  
 17  line_number     1054 non-null   i

In [18]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
top_client5_summarize = top_client5_item_df.groupby("client_id").agg(total_units_purchased = ('qty','sum'),
        total_shipping_price=('shipping_price','sum'),total_revenue=('subtotal','sum'),total_profit=('profit','sum'))

top_client5_summarize.head(10)

total_units_purchased  total_shipping_price  total_revenue  \
client_id                                                               
24741                     239862            5126448.37    70176885.25   
33615                      64313            1828984.89     5839032.11   
38378                      73667            3429455.40     8384321.15   
46820                      75768            1601448.84     7317356.03   
66037                      43018            1395151.85     7995708.38   

           total_profit  
client_id                
24741       36579992.31  
33615        2201994.61  
38378        3271829.89  
46820        2736602.72  
66037        3255031.81

In [21]:
# Format the data and rename the columns to names suitable for presentation.


# Define the money columns. 
top_client5_summarize.rename(columns={"total_units_purchased":"Total Units Purchased", 
                                      "total_revenue":"Total Revenue",
                                      "total_shipping_price":"Total Shipping Price",
                                      "total_profit":"Total Profit",
                                      }, inplace=True)

top_client5_summarize.head()
# Define a function that converts a dollar amount to millions.
def currency_format_millions(x):
    return round(x/1000000,2)

# Apply the currency_format_millions function to only the money columns. 

top_client5_summarize["Total Revenue $(millions)"] = top_client5_summarize["Total Revenue"].apply(currency_format_millions) 
top_client5_summarize["Total Shipping Cost $(millions)"] = top_client5_summarize["Total Shipping Price"].apply(currency_format_millions) 
top_client5_summarize["Total Profit $(millions)"] = top_client5_summarize["Total Profit"].apply(currency_format_millions) 

final_client5_summary = top_client5_summarize[["Total Units Purchased", "Total Shipping Cost $(millions)","Total Revenue $(millions)","Total Profit $(millions)"]]
# Rename the columns to reflect the change in the money format. 
final_client5_summary.head(10)



Total Units Purchased  Total Shipping Cost $(millions)  \
client_id                                                           
24741                     239862                             5.13   
33615                      64313                             1.83   
38378                      73667                             3.43   
46820                      75768                             1.60   
66037                      43018                             1.40   

           Total Revenue $(millions)  Total Profit $(millions)  
client_id                                                       
24741                          70.18                     36.58  
33615                           5.84                      2.20  
38378                           8.38                      3.27  
46820                           7.32                      2.74  
66037                           8.00                      3.26

In [22]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
final_client5_summary_sorted = final_client5_summary.sort_values("Total Profit $(millions)", ascending=False)
final_client5_summary_sorted.head()

Total Units Purchased  Total Shipping Cost $(millions)  \
client_id                                                           
24741                     239862                             5.13   
38378                      73667                             3.43   
66037                      43018                             1.40   
46820                      75768                             1.60   
33615                      64313                             1.83   

           Total Revenue $(millions)  Total Profit $(millions)  
client_id                                                       
24741                          70.18                     36.58  
38378                           8.38                      3.27  
66037                           8.00                      3.26  
46820                           7.32                      2.74  
33615                           5.84                      2.20

In [3]:
print(''' Summary:\n'''
      ''' The client 24741 is major client for company and generates to highest Total Revenue and Total Profit for company.\n'''
      ''' As number of units purchased by client increases, the shipping becomes trivial part of the total price''')

 Summary:
 The client 24741 is major client for company and generates to highest Total Revenue and Total Profit for company.
 As number of units purchased by client increases, the shipping becomes trivial part of the total price
